src/mlpack/methods/reinforcement_learning/replay/prioritized_replay.hpp

```diff
@@ -0,0 +1,212 @@
+   */
+  PrioritizedReplay(const size_t batchSize,
+               const size_t capacity,
+               const double alpha,
+               const size_t dimension = StateType::dimension) :
+      batchSize(batchSize),
+      capacity(capacity),
+      alpha(alpha),
+      position(0),
+      states(dimension, capacity),
+      actions(capacity),
+      rewards(capacity),
+      nextStates(dimension, capacity),
+      isTerminal(capacity),
+      full(false),
+      max_priority(1.0),
+      initial_beta(0.6),
+      replay_beta_iters(10000)
+  { 
+    int size = 1;
+    while (size < capacity) {
+      size *= 2;
+    }
+
+    beta = initial_beta;
+    idxSum = SumTree<double>(size);
+  }
+
+  void Store(const StateType& state,
+             ActionType action,
+             double reward,
+             const StateType& nextState,
+             bool isEnd)
+  {
+    states.col(position) = state.Encode();
+    actions(position) = action;
+    rewards(position) = reward;
+    nextStates.col(position) = nextState.Encode();
+    isTerminal(position) = isEnd;
+
+    idxSum.set(position, max_priority * alpha);
+
+    position++;
+    if (position == capacity)
+    {
+      full = true;
+      position = 0;
+    }
+  }
+
+  arma::ucolvec sampleProportional()
+  {
+    arma::ucolvec idxes(batchSize);
+    double totalSum = idxSum.sum(0, (full ? capacity : position));
+    double sumPerRange = totalSum / batchSize;
+    for (size_t bt = 0; bt < batchSize; bt ++) {
+      double mass = arma::randu() * sumPerRange + bt * sumPerRange;
+      size_t idx = idxSum.findPrefixSum(mass);
+      idxes[bt] = idx;
+    }
+    return idxes;
+  }
+
+  void Sample(arma::mat& sampledStates,
+              arma::icolvec& sampledActions,
+              arma::colvec& sampledRewards,
+              arma::mat& sampledNextStates,
+              arma::icolvec& isTerminal,
+              arma::ucolvec& sampledIndices,
+              arma::rowvec& weights)
+  {
+    size_t upperBound = full ? capacity : position;
+
+    sampledIndices = sampleProportional();
+
+    sampledStates = states.cols(sampledIndices);
+    sampledActions = actions.elem(sampledIndices);
+    sampledRewards = rewards.elem(sampledIndices);
+    sampledNextStates = nextStates.cols(sampledIndices);
+    isTerminal = this->isTerminal.elem(sampledIndices);
+
+//  calculate the weights of sampled transitions
+
+    size_t num_sample = full ? capacity : position;
+    weights = arma::rowvec(sampledIndices.n_rows);
+
+    for (size_t i = 0; i < sampledIndices.n_rows; ++ i)
+    {
+      double p_sample = idxSum.get(sampledIndices[i]) / idxSum.sum();
+      weights[i] = pow(num_sample * p_sample, -beta);
+    }
+    weights /= weights.max();
+  }
+
+  void update_priorities(arma::ucolvec& indices, arma::colvec& priorities)
+  {
+//    update priorities of sampled transitions.
+      for (size_t i = 0; i < indices.n_rows; ++i)
+      {
+        idxSum.set(indices[i], alpha * priorities[i]);
```

Comment:
Hi @robotcator, great work on this :)
Looking this made me wonder if we can improve this step and do the complete update in a single call. So, here's my vague idea:
1. Create a sum tree for `alpha * priorities[i]`, let's name it as priTree.
2. and instead of `idxSum.set(indices[i], alpha * priorities[i]);` for each indices. make a binary call(means traverse each node) to each node of `idxSum` (that will be `O(n)`) and simultaneously update the values of `idxSum` with corresponding value of `priTree`. 
Or since we are making a full update everytime & are just replacing old values with new one, we can just replace the `idxSum` with `priTree`.

It may able to reduce time complexity from `O(nlog(n))` to` O(n)`.

This may not even be a reasonable idea, as it is just a random thought. But do let me know your views on this.

Reply:
Hi, @manish7294, great to see your detailed comments. Your idea is great and seems can reduce time complexity. But I am a little bit confused about the step2. 

The `sumtree` data structure will maintain the `prefix-sum` of samples' priorities,  it is like the [BIT structure](https://en.wikipedia.org/wiki/Fenwick_tree) actually.  Each time we call `idxSum.set(idx, value)`, the corresponding nodes will also be updated. 

From your description, I can not figure out how do you deal with this when you make a binary call( traverse each node) ? Can you give more details?

Reply:
> The sumtree data structure will maintain the prefix-sum of samples' priorities, it is like the BIT structure actually. Though using BIT also could have been a better option.
 
I don't think the implementation is that of a binary indexed tree, it's just like a segment tree.

> Each time we call idxSum.set(idx, value), the corresponding nodes will also be updated.

Right, as we will be traversing the `idxSum` tree we will simultaneously picking up the corresponding values from `priTree` to be used as updates. Moreover since we are doing a complete update of the tree each time, we can even just replace the `idxSum` with `priTree`, I hope this may clear things up a little bit. Let me know if you still have any confusion.

Reply:
> I don't think the implementation is that of a binary indexed tree, it's just like a segment tree.

Yes, you are right. This is my mistake. Both of them can maintain the prefix-sum, but it is segment-tree like data structure here. I add a document for this part. :)   
 
> Right, as we will be traversing the `idxSum` tree we will simultaneously picking up the corresponding values from `priTree` to be used as updates. Moreover since we are doing a complete update of the tree each time, we can even just replace the `idxSum` with `priTree`, I hope this may clear things up a little bit. Let me know if you still have any confusion.

Oh, sounds reasonable. Please let me clarify my understanding to make sure I did not miss what you said. 
1  Instead of calling `idxSum.set` repeatively. We update the leaf node of the sumtree for all the sampled indices.
2 Then update the inner node with bottom-up techniques. 

Oh, yeah, in this way, I think we do not need priTree at all, right?

Reply:
Yes, Exactly, it's basically building a new sum-tree.

Reply:
Hi, @manish7294, I added [`batchUpdate()`](https://github.com/mlpack/mlpack/blob/fc33c5b03ba9c51cc13e91a44baf2ab8db24dedf/src/mlpack/methods/reinforcement_learning/replay/sumtree.hpp#L80) method in `SumTree` class to support batch updating the priorities in `PrioritizedReplay` class.  I also added a simple test case for this method. Please feel free for pointing out if you have any suggestions.